### PR TITLE
Improve errors

### PR DESCRIPTION
**Pull request by erickt**
This PR cleans up the errors to include better information:

* missing metadata or target path is now included in relevant errors.
* unused error variants are removed.
* makes variants for common errors.

### CODE REVIEW

**Review by wellsie1116 — CHANGES_REQUESTED** [review]
errors improved, nice.  Looks like there are some CI failures to address first.

File: tuf/src/error.rs [hunk 1]
```diff
@@ -45,11 +47,35 @@ pub enum Error {
...
+    /// A tempfile persist error occurred.
+    #[error("tempfile: {0}")]
+    TempfilePersist(#[from] tempfile::PersistError),
+
+    /// A tempfile path persist error occurred.
+    #[error("tempfile: {0}")]
+    TempfilePathPersist(#[from] tempfile::PathPersistError),
+
+    /// A derp error occurred.
+    #[error("derp: {0}")]
```
Comment (wellsie1116) [inline tuf/src/error.rs]: "derp" is a cute name, but maybe we should indicate what that means in the error string. "der parse: {0}"?

**Comment by erickt** [comment]
Those pesky features. Should hopefully be fixed now.

File: tuf/src/error.rs [hunk 2]
```diff
@@ -2,18 +2,20 @@
...
+
+/// Alias for `Result<T, Error>`.
+pub type Result<T> = std::result::Result<T, Error>;
 
 /// Error type for all TUF related errors.
 #[non_exhaustive]
 #[derive(Error, Debug)]
 pub enum Error {
     /// The metadata had a bad signature.
     #[error("bad signature")]
```
Comment (wellsie1116) [inline tuf/src/error.rs]: let's include the path in the error string

**Review by wellsie1116 — COMMENTED** [review]
TIL clicking the wrong start review button deletes all my comments

File: tuf/src/error.rs [hunk 3]
```diff
@@ -41,11 +43,35 @@ pub enum Error {
         uri: String,
     },
 
+    /// An IO error occurred.
+    #[error("IO: {0}")]
```
Comment (kellerb) [inline tuf/src/error.rs]: we've been not including source errors in the display text and then using anyhow to stringify the errors for logging, which itself includes the source errors in the stringification

not sure if that pattern would be useful here

File: tuf/src/metadata.rs [hunk 4]
```diff
@@ -2550,11 +2569,17 @@ mod test {
             serde_json::from_value(jsn).unwrap();
         let raw_root = decoded.to_raw().unwrap();
         assert_matches!(
-            verify_signatures(&raw_root, 2, &[root_key.public().clone()]),
-            Err(Error::VerificationFailure(_))
+            verify_signatures(&MetadataPath::root(), &raw_root, 2, &[root_key.public().clone()]),
+            Err(Error::MetadataMissingSignatures { role, number_of_valid_signatures: 1, threshold: 2 })
```
Comment (kellerb) [inline tuf/src/metadata.rs]: some line wraps in the following assert_matches

File: tuf/src/error.rs (hunk 2 above)
Comment (erickt) [reply tuf/src/error.rs]: done

File: tuf/src/metadata.rs (hunk 4 above)
Comment (erickt) [reply tuf/src/metadata.rs]: done

File: tuf/src/error.rs (hunk 1 above)
Comment (erickt) [reply tuf/src/error.rs]: I switched back to the old error writing the error in the `Encoding` variant. I think I want to eventually hide derp properly, but I figure I can do that later.

File: tuf/src/error.rs (hunk 3 above)
Comment (erickt) [reply tuf/src/error.rs]: seems reasonable. done.

**Review by kellerb — APPROVED** [review]

**Pull request merged**